### PR TITLE
[Feature] `WithoutLiersCV` model selection

### PR DESCRIPTION
**Pull request by FBruzzesi**
# Description

Introduces `WithoutLiersCV` as discussed in #307. To be able to follow different cross validation _strategies_, the idea is to take a CV object as input and exclude the anomalous samples from the training indexes. All the splitting logic is delegated to the cv object.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [X] My code follows the style guidelines (flake8)
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (also to the readme.md)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added tests to check whether the new feature adheres to the sklearn convention
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: sklego/model_selection.py [hunk 1]
```diff
@@ -643,3 +643,105 @@ def _regroup(self, groups):
...
+    from sklego.model_selection import WithoutLiersCV
+
+    np.random.seed(1)
+    X = np.random.randn(100, 3)
+    y = np.random.randn(100) > 1.5  # 7% of the data is labeled as anomalous
+
+    cv = WithoutLiersCV(
+        cv=KFold(n_splits=3),
+        anomalous_label=1
+    )
```
Comment (koaning) [inline sklego/model_selection.py:678]: I think I'd want @MBrouns to weight in on the name 😅 just to make sure. 

But I'm also wondering if it's perhaps easier to the enduser to not require an anomalous label ... wouldn't it perhaps be better to pass in an outlier model? this outlier model could then internally train on X and determine which items are outliers. Or am I overthinking?

Comment (FBruzzesi) [reply sklego/model_selection.py:678]: From the conversation in the issue my understanding is slightly different. The goal of the CV is to validate anomaly detectors that do not train with different labels, namely the novelty detection ones. Therefore passing a novelty detection model would not be possible in the first place.

Now I agree that the name would suit both implementations 😁

Comment (FBruzzesi) [reply sklego/model_selection.py:678]: @koaning Potentially we could have two CV strategies:
- `WithoutLiersCV`: takes _any_ outlier detection model, train on `X`, and excludes outliers from `train_indexes`
- `NoveltyDetectorCV`: what's in this PR, to be able to train a novelty detection algorithm on non-anomalous labels and evaluate on both anomalous and not.

**Comment by koaning** [comment]
@FBruzzesi it's been a few years, is this something you still want in? Is this something you'd argue that you have a practical use-case for?

**Comment by FBruzzesi** [comment]
Hey @koaning 👋🏼 

> is this something you still want in?

I don't have a strong opinion to be honest. I am ok if we close this

> Is this something you'd argue that you have a practical use-case for?

_Actually_ I now have a use case for it: as I moved to my current company and we do a bunch of anomaly detection, there are definitely cases in which I check the difference in training with and without anomalous points.

**Comment by koaning** [comment]
If you have a use-case for it, why not add it in?

**Comment by FBruzzesi** [comment]
> If you have a use-case for it, why not add it in?

I got the wrong impression that you just wanted to close this 😅

---

Let me rework this a tiny bit as it's been ~3 years and I learned a few things in the meanwhile. First thing I notice is that the class is not inheriting from anything in scikit, let me double check that

**Comment by koaning** [comment]
I am cool with adding it, just to be clear, but on reflecting I think the state right now is more of a thing of "hey what's the state of the library?". It could be that we take it more in maintenance mode or that we say "nah we should celebrate creativity of components" now that we can have agents help out. 

Either way is totally fine.

**Comment by FBruzzesi** [comment]
> I am cool with adding it, just to be clear, but on reflecting I think the state right now is more of a thing of "hey what's the state of the library?". It could be that we take it more in maintenance mode or that we say "nah we should celebrate creativity of components" now that we can have agents help out.
> 
> Either way is totally fine.

It's true that there has not been much active development, and we are acting in maintenance mode. That said, I would like to keep celebrating the original philosophy of creative freedom and "low stake" experimental components.

**Comment by koaning** [comment]
We always celebrated creative freedom, but we also celebrated the constraint of "should be genuinely useful". As long as both apply, I would love to see more :)

**Comment by FBruzzesi** [comment]
Fully agree Vincent 👌🏼

Anyway, I added a commit (https://github.com/koaning/scikit-lego/pull/595/commits/cb4bd1efb229308474b60e183f311ffbeac3cc93) and PR is ready for review

File: sklego/model_selection.py [hunk 2]
```diff
@@ -902,3 +905,110 @@ def _regroup(self, groups):
         # create a mapper to set every group to the right group_id
         mapper = dict(zip(df["index"], df["group"]))
         return np.vectorize(mapper.get)(groups)
+
+
+class WithoutLiersCV(BaseCrossValidator):
+    r"""Cross-validation splitter that filters out data points labeled as anomalies during the splitting process
+    to exclude them from the training sets.
```
Comment (koaning) [inline sklego/model_selection.py:912]: Do all the outliers appear in the test sets now? I think they don't but it could be good to clarify explicitly.

File: sklego/model_selection.py [hunk 3]
```diff
@@ -902,3 +905,110 @@ def _regroup(self, groups):
...
+        Yields
+        ------
+        train_index : np.ndarray
+            Indices of the training set, excluding anomalous data points.
+        test_index : np.ndarray
+            Indices of the test set, including anomalous data points.
+        """
+        y_arr = np.asarray(y)
+        for train_index, test_index in self.cv.split(X, y, groups):
+            inlier_index = np.where(y_arr[train_index] != self.anomalous_label)[0]
```
Comment (koaning) [inline sklego/model_selection.py]: Does it make sense to have a setting here to also exclude it from the test set?

File: docs/api/model-selection.md [hunk 4]
```diff
@@ -15,6 +15,11 @@
         show_root_full_path: true
         show_root_heading: true
 
+:::sklego.model_selection.WithoutLiersCV
```
Comment (koaning) [inline docs/api/model-selection.md:18]: Maybe we should add a small segment to explain when you'd want this new component? What specific problem it solves. We don't need to have a full dataset/example here but something that makes it feel tangible/plausible of when to use could help people toy with it.

**Review by koaning — CHANGES_REQUESTED** [review]
mostly some small doc suggestions
